staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go

```diff
@@ -0,0 +1,274 @@
+// This can be used to return the canonical form of the CIDR network.
+//
+// - prefixLength: Returns the prefix length of the CIDR in bits.
+// This is the number of bits in the mask.
+//
+// Examples:
+//
+// cidr('192.168.0.0/16').contains(cidr('192.168.10.0/24')) // returns true
+// cidr('192.168.1.0/24').contains(cidr('192.168.2.0/24')) // returns false
+// cidr('192.168.0.0/24').contains(ip('192.168.0.1')) // returns true
+// cidr('192.168.0.0/24').contains(ip('192.168.1.1')) // returns false
+// cidr('192.168.0.0/16').contains('192.168.10.0/24') // returns true
+// cidr('192.168.1.0/24').contains('192.168.2.0/24') // returns false
+// cidr('192.168.0.0/24').contains('192.168.0.1') // returns true
+// cidr('192.168.0.0/24').contains('192.168.1.1') // returns false
+// cidr('192.168.0.1/24').ip() // returns ipAddr('192.168.0.1')
+// cidr('192.168.0.1/24').ip().family() // returns '4'
+// cidr('::1/128').ip() // returns ipAddr('::1')
+// cidr('::1/128').ip().family() // returns '6'
+// cidr('192.168.0.0/24').masked() // returns cidr('192.168.0.0/24')
+// cidr('192.168.0.1/24').masked() // returns cidr('192.168.0.0/24')
+// cidr('192.168.0.0/24') == cidr('192.168.0.0/24').masked() // returns true, CIDR was already in canonical format
+// cidr('192.168.0.1/24') == cidr('192.168.0.1/24').masked() // returns false, CIDR was not in canonical format
+// cidr('192.168.0.0/16').preefixLength() // returns 16
+// cidr('::1/128').prefixLength() // returns 128
+func CIDR() cel.EnvOption {
+	return cel.Lib(cidrsLib)
+}
+
+var cidrsLib = &cidrs{}
+
+type cidrs struct{}
+
+func (*cidrs) LibraryName() string {
+	return "kubernetes.cidr"
+}
+
+var cidrLibraryDecls = map[string][]cel.FunctionOpt{
+	"cidr": {
+		cel.Overload("string_to_cidr", []*cel.Type{cel.StringType}, apiservercel.CIDRType,
+			cel.UnaryBinding(stringToCIDR)),
+	},
+	"contains": {
+		cel.MemberOverload("cidr_contains_string", []*cel.Type{apiservercel.CIDRType, cel.StringType}, cel.BoolType,
+			cel.BinaryBinding(cidrContainsString)),
+		cel.MemberOverload("cidr_contains_ip", []*cel.Type{apiservercel.CIDRType, apiservercel.IPType}, cel.BoolType,
+			cel.BinaryBinding(cidrContainsIP)),
+		cel.MemberOverload("cidr_contains_cidr", []*cel.Type{apiservercel.CIDRType, apiservercel.CIDRType}, cel.BoolType,
+			cel.BinaryBinding(cidrContainsCIDR)),
+	},
+	"ip": {
+		cel.MemberOverload("cidr_ip", []*cel.Type{apiservercel.CIDRType}, apiservercel.IPType,
+			cel.UnaryBinding(cidrToIP)),
+	},
+	"prefixLength": {
+		cel.MemberOverload("cidr_prefix_length", []*cel.Type{apiservercel.CIDRType}, cel.IntType,
+			cel.UnaryBinding(prefixLength)),
+	},
+	"masked": {
+		cel.MemberOverload("cidr_masked", []*cel.Type{apiservercel.CIDRType}, apiservercel.CIDRType,
+			cel.UnaryBinding(masked)),
+	},
+	"isCIDR": {
+		cel.Overload("is_cidr", []*cel.Type{cel.StringType}, cel.BoolType,
+			cel.UnaryBinding(isCIDR)),
+	},
+}
+
+func (*cidrs) CompileOptions() []cel.EnvOption {
+	options := []cel.EnvOption{}
```

Comment:
Don't forget to add `cel.Types(apiservercel.IPType, apiservercel.CIDRType)` as an environment option

Reply:
I don't actually know why we need that here, can you expand on why those are needed?

Reply:
These are type identifiers which are needed if the type is referenced directly or as a field in a struct. If you don't provide the identifier expressions like the following should fail: `type(<ip>) == net.IP`, and references to ip types within fields won't resolve correctly.

Reply:
I tried to add a test `type(ip('192.168.0.1')) == net.IP` and I'm getting `"undeclared reference to 'net' (in container '')"`. I can see that the net.IP is registered in `revTypeMap` but when I dive deeper into the code, it's not registered as an ident, which is where it's being looked up. Any thoughts?

Reply:
@JoelSpeed I think there might be a slight bug in the way top-level types are resolved during type-check. As a workaround, include the following line after you add the type: 

```
cel.Variable(apiservercel.IPType.TypeName(), cel.NewTypeTypeWithParam(apiservercel.IPType))
```

I've tested the fix within CEL and even if you leave the workaround in, it will just become a noop once the fix is made. I'll follow up with a CEL PR reference.